File: 070_ImplementTrie.py
"""
Topic: Design / Trie
Problem: Implement Trie (Prefix Tree)
Description: Implement insert, search, and startsWith.
Time: insert/search/startsWith O(L) where L is word length
Space: O(total characters)
"""

from typing import Dict


class TrieNode:
    def __init__(self):
        self.children: Dict[str, 'TrieNode'] = {}
        self.is_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.is_word = True

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return node.is_word

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return True


if __name__ == '__main__':
    trie = Trie()
    trie.insert("apple")
    print(trie.search("apple"))  # True
    print(trie.search("app"))    # False
    print(trie.startsWith("app"))  # True
    trie.insert("app")
    print(trie.search("app"))    # True

File: 070_ImplementTrie.cpp
// Topic: Design / Trie
// Implement Trie (Prefix tree) with insert, search, startsWith
// Time Complexity: insert/search O(L) where L is word length
// Space Complexity: O(total chars)

#include <bits/stdc++.h>
using namespace std;

struct TrieNode
{
    bool end = false;
    array<TrieNode*, 26> nxt{};
    TrieNode() { nxt.fill(nullptr); }
};

class Trie
{
    TrieNode* root;

   public:
    Trie() : root(new TrieNode()) {}
    void insert(const string& w)
    {
        TrieNode* p = root;
        for (char c : w)
        {
            int i = c - 'a';
            if (!p->nxt[i])
                p->nxt[i] = new TrieNode();
            p = p->nxt[i];
        }
        p->end = true;
    }
    bool search(const string& w)
    {
        TrieNode* p = root;
        for (char c : w)
        {
            int i = c - 'a';
            if (!p->nxt[i])
                return false;
            p = p->nxt[i];
        }
        return p->end;
    }
    bool startsWith(const string& pref)
    {
        TrieNode* p = root;
        for (char c : pref)
        {
            int i = c - 'a';
            if (!p->nxt[i])
                return false;
            p = p->nxt[i];
        }
        return true;
    }
};

int main()
{
    Trie trie;
    trie.insert("apple");
    cout << trie.search("apple") << "\n";    // 1
    cout << trie.search("app") << "\n";      // 0
    cout << trie.startsWith("app") << "\n";  // 1
    trie.insert("app");
    cout << trie.search("app") << "\n";  // 1
    return 0;
}
